### ISSUE TITLE
Aux loss increases after a couple thousand iterations

### ISSUE THREAD

**Issue by makinyilmaz** [InterDigitalInc/CompressAI#148]
First of all I want to thank you for your precious work ! 

I have a question about aux loss. I am trying to train a video compression framework where I have a flow compressor and a residual compressor. I initialized optimizer and aux optimizer as you stated under examples directory. I set aux learning rate to 1.e-3 and it first decreases as training continues (I sum flow and residual aux losses). However at some point, it starts to increase. I tried to decrease the aux learning rate to 1.e-5 after enough iterations. Aux loss first decreases but bounces back and starts to increase. Do you have any idea what could cause this observation?

Thanks a lot

**Comment by ildmxGL** [comment]
I am in similar problem. The aux loss of EntropyBottleneck is not converge to 0. It takes 10 and converges at 50. Is there any training strategies for the best results?

Thanks for your works.

**Comment by fracape** [comment]
Hi, 
sorry @makinyilmaz for the late reply. As you saw, the auxiliary loss aims at refining the boundaries between which the range coder will be used, i.e. the boundaries of the CDFs that will be used for the actual entropy coding using a range coder, e.g. rANS. We followed the initial tensorflow compression training strategy which learned both entropy parameters and the rest during the whole training process. The training loop contained in Compressai is a minimalist template trainer, please feel free to try other strategies. We can compare and discuss observations in the Discussion sections.

**Comment by makinyilmaz** [comment]
Hi @fracape thanks for your reply. I tried applying a seperate gradient clipping on aux_parameters in addition to network parameters and observed the aux loss is converged. Dear @ildmxGL, you may try a similar strategy.

**Comment by xkyi1111** [comment]
Hi @makinyilmaz can you share how to implement the seperate gradient clipping on aux_parameters ?
Here is my code, it does not work.
`            out_criterion['loss'].backward()
            torch.nn.utils.clip_grad_norm_((params_dict[n] for n in sorted(parameters)), 1)
            optimizer.step()

            aux_loss = compute_aux_loss(model, accelerator, backward=True)
            torch.nn.utils.clip_grad_norm_((params_dict[n] for n in sorted(aux_parameters)), 1)
            aux_optimizer.step()`

**Comment by xkyi1111** [comment]
> Hi @fracape thanks for your reply. I tried applying a seperate gradient clipping on aux_parameters in addition to network parameters and observed the aux loss is converged. Dear @ildmxGL, you may try a similar strategy.

Hi @makinyilmaz can you share how to implement the seperate gradient clipping on aux_parameters ?
Here is my code, it does not work.

            out_criterion['loss'].backward()
            torch.nn.utils.clip_grad_norm_((params_dict[n] for n in sorted(parameters)), 1)
            optimizer.step()

            aux_loss = compute_aux_loss(model, accelerator, backward=True)
            torch.nn.utils.clip_grad_norm_((params_dict[n] for n in sorted(aux_parameters)), 1)
            aux_optimizer.step()

**Comment by YodaEmbedding** [comment]
@xkyi1111 My guess is that you may need to clip to a smaller value, e.g. 0.1 or 0.01. Perhaps:

```
print(norm(aux_loss.grad())) # determine reasonable value

if aux_loss < 100:
    grad_clip(..., reasonable_value)
```

**Comment by xkyi1111** [comment]
> @xkyi1111 My guess is that you may need to clip to a smaller value, e.g. 0.1 or 0.01. Perhaps:
> 
> ```
> print(norm(aux_loss.grad())) # determine reasonable value
> 
> if aux_loss < 100:
>     grad_clip(..., reasonable_value)
> ```
> 
> Also note that beyond a certain point, minimizing `aux_loss` doesn't have too much effect on entropy coder performance.

@YodaEmbedding thanks for your reply！I have tried to clip it to a smaller value and the result is unchanged. Besides, aux_loss.grad() is None. In my code, after a few thousand iterations, the aux loss and mse loss became nan, which I thought was caused by the increase in aux loss. Is this possible？

**Comment by YodaEmbedding** [comment]
Is `aux_loss.backward()` called? 

Also, I meant to say: try checking the grad norm of the aux_parameters gradients.

**Comment by xkyi1111** [comment]
> What value of `aux_loss` do you get after the zeroth, first, and last epochs?
> 
> Sorry, I meant to say: try checking the grad norm of the aux_parameters gradients.

Sorry, I'm still in the development phase of my code, the previous nan results were not saved. I am retraining the code. Besides, I use [accelerate ](https://github.com/huggingface/accelerate)to implement ddp and amp, I am not sure if it will cause nan result. I will give feedback if there is any news.

**Comment by xkyi1111** [comment]
Hi @YodaEmbedding, here is my training log, are there any good suggestions?

```
01/02/2023 05:22:21 - Epoch [0/200], Step [1500/4039]: [ avg_rd: 14.47 | cur_rd: 13.22 | avg_mse: 0.01 | cur_mse: 0.01 | avg_aux: 5059.33 | cur_aux: 4696.83 | avg_bpp: 0.3884 | cur_bpp: 0.3296 | avg_psnr: 21.63 | cur_psnr: 22.01 | iter_time: 1.15 s/it | time_left: 00:36:42 | ela_time: 3982]
01/02/2023 05:22:42 - Epoch [0/200], Step [1510/4039]: [ avg_rd: 14.43 | cur_rd: 3.02 | avg_mse: 0.01 | cur_mse: 0.00 | avg_aux: 5056.89 | cur_aux: 4684.98 | avg_bpp: 0.3878 | cur_bpp: 0.2757 | avg_psnr: 21.64 | cur_psnr: 28.73 | iter_time: 1.21 s/it | time_left: 00:34:47 | ela_time: 4003]
01/02/2023 05:23:21 - Epoch [0/200], Step [1520/4039]: [ avg_rd: nan | cur_rd: nan | avg_mse: nan | cur_mse: nan | avg_aux: nan | cur_aux: nan | avg_bpp: nan | cur_bpp: nan | avg_psnr: nan | cur_psnr: nan | iter_time: 1.19 s/it | time_left: 00:35:15 | ela_time: 4042]
01/02/2023 05:23:59 - Epoch [0/200], Step [1530/4039]: [ avg_rd: nan | cur_rd: nan | avg_mse: nan | cur_mse: nan | avg_aux: nan | cur_aux: nan | avg_bpp: nan | cur_bpp: nan | avg_psnr: nan | cur_psnr: nan | iter_time: 1.28 s/it | time_left: 00:32:41 | ela_time: 4080]
01/02/2023 05:24:20 - Epoch [0/200], Step [1540/4039]: [ avg_rd: nan | cur_rd: nan | avg_mse: nan | cur_mse: nan | avg_aux: nan | cur_aux: nan | avg_bpp: nan | cur_bpp: nan | avg_psnr: nan | cur_psnr: nan | iter_time: 1.16 s/it | time_left: 00:35:47 | ela_time: 4101]
```
